ISPN-13355 Support Remote LDAP Server for testing purposes

diegolovison opened the pull request:
https://issues.redhat.com/browse/ISPN-13355

diegolovison commented on server/tests/src/test/java/org/infinispan/server/test/core/ldap/README.md:
````diff
@@ -0,0 +1,20 @@
+mvn verify -s maven-settings.xml -pl server/tests -Dit.test="AuthorizationLDAPIT,AuthenticationImplicitIT,AuthenticationLDAPIT,AuthenticationMultiEndpointIT" -Dorg.infinispan.test.ldapServer=empty -fae
+```
+```
+[ERROR] Tests run: 202, Failures: 14, Errors: 29, Skipped: 0
````
@tristantarrant why when using an LDAP server that does nothing. Do we still have test passing?
For the tests above `AuthorizationLDAPIT,AuthenticationImplicitIT,AuthenticationLDAPIT,AuthenticationMultiEndpointIT`, Do we still need the LDAP server?

tristantarrant commented:
I ran the tests with `mvn verify -pl server/tests -Dit.test=AuthenticationLDAPIT -Dorg.infinispan.test.ldapServer=empty`

and I got:

```
[INFO] Results:
[INFO] 
[ERROR] Errors: 
[ERROR]   AuthenticationLDAPIT.testReadWrite[DIGEST-MD5] » Transport org.infinispan.clie...
[ERROR]   AuthenticationLDAPIT.testReadWrite[DIGEST-SHA-256] » Transport org.infinispan....
[ERROR]   AuthenticationLDAPIT.testReadWrite[DIGEST-SHA-384] » Transport org.infinispan....
[ERROR]   AuthenticationLDAPIT.testReadWrite[DIGEST-SHA-512] » Transport org.infinispan....
[ERROR]   AuthenticationLDAPIT.testReadWrite[DIGEST-SHA] » Transport org.infinispan.clie...
[ERROR]   AuthenticationLDAPIT.testReadWrite[PLAIN] » Transport org.infinispan.client.ho...
[ERROR]   AuthenticationLDAPIT.testReadWrite[SCRAM-SHA-1] » Transport org.infinispan.cli...
[ERROR]   AuthenticationLDAPIT.testReadWrite[SCRAM-SHA-256] » Transport org.infinispan.c...
[ERROR]   AuthenticationLDAPIT.testReadWrite[SCRAM-SHA-384] » Transport org.infinispan.c...
[ERROR]   AuthenticationLDAPIT.testReadWrite[SCRAM-SHA-512] » Transport org.infinispan.c...
[INFO] 
[ERROR] Tests run: 11, Failures: 0, Errors: 10, Skipped: 0
```

tristantarrant commented:
AuthorizationLDAPIT gives me:
```
AuthorizationLDAPIT.testAdminAndDeployerCanManageSchema » Transport org.infini...
[ERROR]   AuthorizationLDAPIT.testAdminCanRemoveCacheWithoutRole » Transport org.infinis...
[ERROR]   AuthorizationLDAPIT.testBulkReadUsersCanQuery » Transport org.infinispan.clien...
[ERROR]   AuthorizationLDAPIT.testExecScripts » Transport org.infinispan.client.hotrod.e...
[ERROR]   AuthorizationLDAPIT.testHotRodAdminAndDeployerCanDoEverything » Transport org....
[ERROR]   AuthorizationLDAPIT.testHotRodBulkOperationsExplicit » Transport org.infinispa...
[ERROR]   AuthorizationLDAPIT.testHotRodBulkOperationsImplicit » Transport org.infinispa...
[ERROR]   AuthorizationLDAPIT.testHotRodReaderCannotWriteExplicit » Transport org.infini...
[ERROR]   AuthorizationLDAPIT.testHotRodReaderCannotWriteImplicit » Transport org.infini...
[ERROR]   AuthorizationLDAPIT.testHotRodWriterCannotReadExplicit » Transport org.infinis...
[ERROR]   AuthorizationLDAPIT.testHotRodWriterCannotReadImplicit » Transport org.infinis...
[ERROR]   AuthorizationLDAPIT.testNonBulkReadUsersCannotQuery » Transport org.infinispan...
[ERROR]   AuthorizationLDAPIT.testNonCreatorsSchema » Transport org.infinispan.client.ho...
[ERROR]   AuthorizationLDAPIT>AbstractAuthorization.testRestAdminCanDoEverything:115 » Security
[ERROR]   AuthorizationLDAPIT>AbstractAuthorization.testRestReaderCannotWriteExplicit:260->AbstractAuthorization.testRestReaderCannotWrite:264->AbstractAuthorization.restCreateAuthzCache:586 » Security
[ERROR]   AuthorizationLDAPIT>AbstractAuthorization.testRestReaderCannotWriteImplicit:255->AbstractAuthorization.testRestReaderCannotWrite:264->AbstractAuthorization.restCreateAuthzCache:586 » Security
[ERROR]   AuthorizationLDAPIT>AbstractAuthorization.testRestWriterCannotReadExplicit:217->AbstractAuthorization.testRestWriterCannotRead:221->AbstractAuthorization.restCreateAuthzCache:586 » Security
[ERROR]   AuthorizationLDAPIT>AbstractAuthorization.testRestWriterCannotReadImplicit:171->AbstractAuthorization.testRestWriterCannotRead:221->AbstractAuthorization.restCreateAuthzCache:586 » Security
[ERROR]   AuthorizationLDAPIT.testScriptUpload » Transport org.infinispan.client.hotrod....
```

tristantarrant commented:
AuthenticationImplicitIT and AuthenticationMultiEndpointIT have nothing to do with LDAP

diegolovison commented:
@tristantarrant in this case:
* AuthenticationImplicitIT: can I remove `.addListener(new LdapServerListener())` ?
* AuthenticationImplicitIT: can I remove `.addListener(new LdapServerListener())` ?

I am assuming that all tests that contain `LdapServerListener` has a relationship with LDAP

tristantarrant commented:
Those were probably leftovers

diegolovison commented:
@tristantarrant I have updated the PR

Few things to have in mind:
* Removed `LdapServerListener` from `AuthenticationImplicitIT` and `AuthenticationMultiEndpointIT`
* Moved `dn` from `infinispan-kerberos.ldif` and `infinispan.ldif` to `infinispan-dn.ldif`
* `infinispan-dn.ldif` is executed only for Apache LDAP Server. Keycloak Ldap has a default `dn`. I wasn't able to change it. This is the reason for the split.
* `ApacheLdapServer` was copied from the original code.
* `LdapServerListener` becomes a delegate to an LDAP Server

It is working:
* Debug `org.infinispan.server.security.authorization.AuthorizationLDAPIT#testBulkReadUsersCanQuery` without any additional VM argument. It will execute with the Apache Server and the test will pass.

It is not working
* Debug `org.infinispan.server.security.authorization.AuthorizationLDAPIT#testBulkReadUsersCanQuery` with `-Dorg.infinispan.test.ldapServer=rhds10`. The value of `identity.getEntry().getAttributes()` is `empty` in `org.wildfly.security.auth.realm.ldap.LdapSecurityRealm::LdapRealmIdentity::getCredential` when `"clear".equals(algorithmName)`. The line is https://github.com/wildfly-security/wildfly-elytron/blob/4fff4dc05c82856438d4a42aae6686015a4f0df7/auth/realm/ldap/src/main/java/org/wildfly/security/auth/realm/ldap/LdapSecurityRealm.java#L426

I was not able to understand why

tristantarrant commented:
It may depend on how the user password attribute is configured on RHDS. It may be necessary to use `direct-verification` instead.

tristantarrant commented:
Cherry-picked to 13.0.x too. Thanks @diegolovison
